wg-api/spec-documents/clipboard-rearchitecture.md

```diff
@@ -0,0 +1,105 @@
+- [#11838](https://github.com/electron/electron/issues/11838)
+- [#9035](https://github.com/electron/electron/issues/9035)
+- [#5078](https://github.com/electron/electron/issues/5078)
+- [electron/governance#229](https://github.com/electron/governance/pull/229)
+
+Electron provided a `clipboard` API to its users preceding the specification of the [Clipboard API by the W3C](https://w3c.github.io/clipboard-apis/#clipboard-interface). As a result, we locked into an API surface and created an API contract with our users that became progressively harder to alter without breaking said contract while the web drove a more standardized approach. On top of this confusion, the `clipboard` module provided by Electron is a dual-process module, meaning that it exists in parallel to the more robust API provided by Chromium in the renderer process via `navigator.clipboard` global. 
+
+At this point, it becomes clear that it is in users' and maintainers' best interest to reconsider how we approach the `clipboard` module - breaking changes might be difficult in the short term but provide more consistency and functionality as we look to the future of the module.
+
+## API Design
+
+The largest consideration for changes to the `clipboard` module is to what degree we should choose to align with the W3C specification. As we consider this, we should consider what needs or use cases might exist for Electron uses which necessarily wouldn't exist for Chrome users, and how we would plan to account for that in our new design.
+
+### Potential Options
+
+1. Strict alignment with W3C specification - no additional APIs exposed.
+2. Alignment with W3C where possible, with some additional APIs exposed to handle desktop use cases which aren’t addressed by the specification.
+3. Leave API as-is and continue to adapt it as required by upstream changes by Chromium development.
+ 
+### Preferred Approach/Solution
+
+Alignment with W3C where possible, with some additional APIs exposed to handle desktop use cases which aren’t considered by the specification.
+
+**APIs to Preserve**
+* `clipboard.readBookmark()`
+  * Not supported through existing Web APIs or in the specification.
+* `clipboard.writeBookmark(title, url[, type])`
+  * Not supported through existing Web APIs or in the specification.
+* `clipboard.readFindText()` _macOS_
+  * Not supported through existing Web APIs or in the specification.
+* `clipboard.writeFindText(text)`
+  * Not supported through existing Web APIs or in the specification.
+* `clipboard.clear([type])`
+  * Not supported through existing Web APIs or in the specification.
+* `clipboard.readText([type])`
+  * Already implemented per specification.
+* `clipboard.writeText(text[, type])`
+  * Already implemented per specification.
+
+**APIs to Remove**
+* `clipboard.availableFormats([type])`
+  * Superseded by `clipboard.read` [Web API](https://developer.mozilla.org/en-US/docs/Web/API/Clipboard/read)
+* `clipboard.has(format[, type])` 
```

Comment:
Having a `has` API is probably still needed, otherwise one has to use `read` to get all types of data at once, and the literate them to find the needed one.

Reply:
`has` can be replicated with a fairly short amount of JS:
```js
async function has(format, clipboardType) {
  const clipboardItems = await clipboard.read(clipboardType);
  const matchingClipboardItem = clipboardItems.find(clipboardItem => {
    return clipboardItem.types.includes(format);
  });
  if (matchingClipboardItem) {
    return true;
  }
}
```

Reply:
I'd be in favor of keeping `has` as well - I'd probably also be in favor of leaving it in cpp given the vast bulk of clipboard will be implemented at that layer and it's likely marginally faster.